setPageLayout props type rejects arrays of plain objects despite being JSON-serializable

### Environment

- Operating System: Darwin (macOS 25.3.0, arm64)
- Node Version: v25.7.0 (local) / v22 (StackBlitz — also reproduces)
- Package Manager: pnpm@10.30.0
- Nuxt Version: 4.4.2
- Nitro Version: 4.4.2 (@nuxt/nitro-server)
- Vue Version: 3.5.33
- Vue Router Version: 5.0.5
- Vite Version: 7.3.2
- Bundler: @nuxt/vite-bundler 4.4.2
- Nuxt CLI Version: 3.34.0
- User Config: compatibilityDate, devtools
- Runtime Modules: -
- Build Modules: -
- Runtime Modules: —
- Build Modules: —

### Reproduction

- Repo: https://github.com/mew-ton/__nuxt-setpagelayout-rejects-object-arrays-repro
- StackBlitz: https://stackblitz.com/github/mew-ton/__nuxt-setpagelayout-rejects-object-arrays-repro?file=app%2Fpages%2Fdocuments%2Fresume.vue

Steps:

1. `pnpm install`
2. `pnpm type-check`
3. Observe four type errors — three in the realistic breadcrumbs pages, and one in a minimal `{ label: string; count: number }[]` page with no generics and no imported types.


### Describe the bug


`setPageLayout(name, props)` — introduced in [#33805](https://github.com/nuxt/nuxt/pull/33805) — advertises JSON-serializable values as valid props, but the actual type helper `MakeSerializableObject` rejects **any array whose elements are not primitives**, even when the array is trivially JSON-serializable.

A very common real-world use case is passing a breadcrumbs array to a shared default layout:

```ts
setPageLayout("default", {
  breadcrumbs: [
    { title: "Top", to: "/" },
    { title: "Documents", to: "/documents" },
    { title: "My Resume" },
  ],
});
```

This value:

- Round-trips losslessly through `JSON.stringify` / `JSON.parse`.
- Is delivered to the layout correctly at runtime — the breadcrumbs strip renders as expected.
- Is the natural Vue/Nuxt shape: every item carries a `to` compatible with `<NuxtLink>`.

Yet TypeScript rejects it with `TS2322: Type '…[]' is not assignable to type 'undefined'`.

The bug is **not** specific to `RouteLocationRaw` or to complex / imported types. Even a trivial `{ label: string; count: number }[]` is rejected (see `app/pages/simple.vue` in the repro — no generics, no imported types, just plain JSON-serializable objects).

### Inconsistency with single objects

A single object of the same shape is accepted — only the array form fails:

```ts
setPageLayout("default", { item: { title: "Top" } })              // ✅ OK
setPageLayout("default", { breadcrumbs: [{ title: "Top" }] })     // ❌ never
```

Root cause:

`MakeSerializableObject` in `packages/nuxt/src/pages/runtime/utils.ts` (also visible in `node_modules/nuxt/dist/pages/runtime/utils.d.ts` of 4.4.2):

```ts
export type MakeSerializableObject<T> = T extends Function | symbol
  ? never
  : {
      [K in keyof T]: T[K] extends SerializablePrimitive
        ? T[K]
        : T[K] extends (infer U)[]
          ? U extends SerializablePrimitive
            ? T[K]
            : never   // ← rejects every array of non-primitives
          : T[K] extends Record<string, unknown>
            ? T[K]
            : never;
    };
```

The array branch only admits arrays of primitives, which is strictly narrower than the stated "JSON-serializable" contract. Note the sibling branch for single objects already accepts `Record<string, unknown>`, so the asymmetry is the source of the confusion.

Suggested fix:

Model the constraint with a recursive JSON-value type that matches the actual serialization contract:

```ts
type JSONPrimitive = string | number | boolean | null;
type JSONValue = JSONPrimitive | JSONValue[] | { [k: string]: JSONValue };
```

At minimum, the array branch should accept arrays of `Record<string, unknown>` (mirroring the existing single-object branch):

```ts
T[K] extends (infer U)[]
  ? U extends SerializablePrimitive | Record<string, unknown>
    ? T[K]
    : never
  : ...
```

### Additional context

- The runtime path works: the breadcrumbs strip in `app/layouts/default.vue` renders correctly for all four reproduction pages. The bug is strictly in the type definition.
- Four reproduction pages in the repo cover: 1-item (no `to`), 2-item (string `to`), 3-item (object-form `to` with query params), and a **minimal** case with plain `{ label: string; count: number }[]` and no imported types / no generics.


### Logs

```shell-script
$ pnpm type-check

> nuxt-setpagelayout-rejects-object-arrays-repro@ type-check
> nuxt typecheck

app/pages/documents/index.vue(6,3): error TS2322: Type '({ title: string; to: string; } | { title: string; })[]' is not assignable to type 'undefined'.
app/pages/documents/resume.vue(6,3): error TS2322: Type '({ title: string; to: string; } | { title: string; to: { path: string; query: { sort: string; }; }; } | { title: string; })[]' is not assignable to type 'undefined'.
app/pages/index.vue(9,3): error TS2322: Type '{ title: string; }[]' is not assignable to type 'undefined'.
app/pages/simple.vue(14,3): error TS2353: Object literal may only specify known properties, and 'items' does not exist in type '{ readonly breadcrumbs?: undefined; key?: undefined; ref?: undefined; ref_for?: boolean | undefined; ref_key?: string | undefined; onVnodeBeforeMount?: undefined; onVnodeMounted?: undefined; ... 5 more ...; style?: undefined; }'.
```

--- a/packages/nuxt/src/pages/runtime/utils.ts
+++ b/packages/nuxt/src/pages/runtime/utils.ts
@@ -10,17 +10,17 @@ type SerializablePrimitive = string | number | boolean | null | undefined
 /** JSON-serializable value (non-recursive definition to avoid excessive type depth) */
 export type SerializableValue = SerializablePrimitive | SerializablePrimitive[] | Record<string, unknown>
 
+export type MaybeArray<T> = T | T[]
+
 /** Constrains T to only contain serializable properties. Non-serializable properties become `never`. */
 // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
-export type MakeSerializableObject<T> = T extends Function | symbol ? never : {
-  [K in keyof T]: T[K] extends SerializablePrimitive
-    ? T[K]
-    : T[K] extends (infer U)[]
-      ? U extends SerializablePrimitive ? T[K] : never
-      : T[K] extends Record<string, unknown>
+export type MakeSerializableObject<T> = T extends Function | symbol
+  ? never
+  : {
+      [K in keyof T]: T[K] extends MaybeArray<SerializablePrimitive | Record<string, unknown>>
         ? T[K]
         : never
-}
+    }
 
 const ROUTE_KEY_PARENTHESES_RE = /(:\w+)\([^)]+\)/g
 const ROUTE_KEY_SYMBOLS_RE = /(:\w+)[?+*]/g

--- a/packages/nuxt/src/pages/runtime/utils.ts
+++ b/packages/nuxt/src/pages/runtime/utils.ts
@@ -10,14 +10,14 @@ type SerializablePrimitive = string | number | boolean | null | undefined
 export type MaybeArray<T> = T | T[]
 
 /** JSON-serializable value (non-recursive definition to avoid excessive type depth) */
-export type SerializableValue = MaybeArray<SerializablePrimitive | Record<string, unknown>>
+export type SerializableValue = MaybeArray<SerializablePrimitive | Record<string, SerializablePrimitive>>
 
 /** Constrains T to only contain serializable properties. Non-serializable properties become `never`. */
 // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
 export type MakeSerializableObject<T> = T extends Function | symbol
   ? never
   : {
-      [K in keyof T]: T[K] extends MaybeArray<SerializablePrimitive | Record<string, unknown>>
+      [K in keyof T]: T[K] extends SerializableValue
         ? T[K]
         : never
     }

--- a/packages/nuxt/src/pages/runtime/utils.ts
+++ b/packages/nuxt/src/pages/runtime/utils.ts
@@ -7,10 +7,10 @@ export type RouterViewSlotProps = Parameters<RouterViewSlot>[0]
 
 type SerializablePrimitive = string | number | boolean | null | undefined
 
-/** JSON-serializable value (non-recursive definition to avoid excessive type depth) */
-export type SerializableValue = SerializablePrimitive | SerializablePrimitive[] | Record<string, unknown>
-
 export type MaybeArray<T> = T | T[]
+
+/** JSON-serializable value (non-recursive definition to avoid excessive type depth) */
+export type SerializableValue = MaybeArray<SerializablePrimitive | Record<string, unknown>>
 
 /** Constrains T to only contain serializable properties. Non-serializable properties become `never`. */
 // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type

